Hessian-vector product of the residual with AD

kliegeois opened the pull request:
This PR adds the capability to use Automatic Differentiation (AD) to compute the Hessian-vector products of the residual by nesting AD types.

This PR adds four contributions of the Hessian-vector products to the previously merged ones:
- <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{H}_{\boldsymbol{x}\boldsymbol{x}}\boldsymbol{v}_{\boldsymbol{x}}=\left.\frac{\partial}{\partial r} \nabla_{\boldsymbol{x}} \left\langle \boldsymbol{f}(\boldsymbol{x}%2Br \boldsymbol{v}_{\boldsymbol{x}},\boldsymbol{p}_1, \boldsymbol{p}_2),\boldsymbol{z}\right\rangle\right|_{r=0}">,
- <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{H}_{\boldsymbol{x}\boldsymbol{p}_1}\boldsymbol{v}_{\boldsymbol{p}_1}=\left.\frac{\partial}{\partial r} \nabla_{\boldsymbol{x}} \left\langle \boldsymbol{f}(\boldsymbol{x},\boldsymbol{p}_1%2Br \boldsymbol{v}_{\boldsymbol{p}_1}, \boldsymbol{p}_2),\boldsymbol{z}\right\rangle\right|_{r=0}">,
- <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{H}_{\boldsymbol{p}_2\boldsymbol{x}}\boldsymbol{v}_{\boldsymbol{x}}=\left.\frac{\partial}{\partial r} \nabla_{\boldsymbol{p}_2} \left\langle \boldsymbol{f}(\boldsymbol{x}%2Br \boldsymbol{v}_{\boldsymbol{x}},\boldsymbol{p}_1, \boldsymbol{p}_2),\boldsymbol{z}\right\rangle\right|_{r=0}">,
- <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{H}_{\boldsymbol{p}_2\boldsymbol{p}_1}\boldsymbol{v}_{\boldsymbol{p}_1}=\left.\frac{\partial}{\partial r} \nabla_{\boldsymbol{p}_2} \left\langle \boldsymbol{f}(\boldsymbol{x},\boldsymbol{p}_1%2Br \boldsymbol{v}_{\boldsymbol{p}_1}, \boldsymbol{p}_2),\boldsymbol{z}\right\rangle\right|_{r=0}">, 

where <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{x}"> is the solution, <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{p}_1"> is a first distributed parameter, <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{p}_2"> is a potentially different second distributed parameter, <img src="https://render.githubusercontent.com/render/math?math=\large \boldsymbol{f}"> is the residual, <img src="https://render.githubusercontent.com/render/math?math=\large \boldsymbol{z}"> is the Lagrange multiplier vector, <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{v}_\boldsymbol{x}"> is a direction vector with the same dimension as the vector <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{x}">, <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{v}_{\boldsymbol{p}_1}"> is a direction vector with the same dimension as the vector <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{p}_1">, and <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{v}_{\boldsymbol{p}_2}"> is a direction vector with the same dimension as the vector <img src="https://render.githubusercontent.com/render/math?math=\large\boldsymbol{p}_2">.

The implementation has been tested comparing results computed with AD and with Finite Differences (FD) of the gradients on a heat problem and FO_GIS examples.
However, this PR does not include tests yet. They will be included in a second PR.

mperego commented on src/LandIce/evaluators/LandIce_Gather2DField_Def.hpp:
```diff
@@ -357,8 +363,14 @@ evaluateFields(typename Traits::EvalData workset)
+  bool f_xp_is_active = !workset.hessianWorkset.hess_vec_prod_f_xp.is_null();
+  bool f_px_is_active = !workset.hessianWorkset.hess_vec_prod_f_px.is_null();
+
+  const bool is_active = g_xx_is_active || g_xp_is_active || f_xx_is_active || f_xp_is_active;
```
Can you add a comment explaining what `is_active` and `is_direction_active` are?

mperego commented on src/LandIce/evaluators/LandIce_GatherVerticallyContractedSolution_Def.hpp:
```diff
@@ -416,8 +416,14 @@ evaluateFields(typename Traits::EvalData workset)
+  bool f_px_is_active = !workset.hessianWorkset.hess_vec_prod_f_px.is_null();
 
-  if(g_xx_is_active||g_px_is_active) {
+  const bool is_active = g_xx_is_active || g_xp_is_active || f_xx_is_active || f_xp_is_active;
```
same here. Can you add a comment explaining what these variables are?

mperego commented on src/evaluators/bc/PHAL_Dirichlet.hpp:
```diff
@@ -120,6 +120,7 @@ class Dirichlet<PHAL::AlbanyTraits::HessianVec,Traits>
    : public DirichletBase<PHAL::AlbanyTraits::HessianVec, Traits> {
 public:
   Dirichlet(Teuchos::ParameterList& p);
+  void preEvaluate(typename Traits::EvalData d);
```
please add a comment briefly explaining why we need pre-evaluate.

mperego commented on src/evaluators/bc/PHAL_DirichletField.hpp:
```diff
@@ -102,6 +102,7 @@ class DirichletField<PHAL::AlbanyTraits::HessianVec, Traits>
   public:
     DirichletField(Teuchos::ParameterList& p);
     typedef typename PHAL::AlbanyTraits::HessianVec::ScalarT ScalarT;
+    void preEvaluate(typename Traits::EvalData d);
```
please add a comment briefly explaining why we need preEvaluate.

mperego commented on src/evaluators/bc/PHAL_SDirichlet.hpp:
```diff
@@ -124,6 +124,9 @@ class SDirichlet<PHAL::AlbanyTraits::HessianVec, Traits>
   SDirichlet(Teuchos::ParameterList& p);
 
+  void
+  preEvaluate(typename Traits::EvalData d);
```
please add a comment briefly explaining why we need preEvaluate.

mperego commented on src/evaluators/bc/PHAL_SDirichletField.hpp:
```diff
@@ -127,6 +127,7 @@ class SDirichletField<PHAL::AlbanyTraits::HessianVec, Traits>
 
     SDirichletField(Teuchos::ParameterList& p);
 
+    void preEvaluate(typename Traits::EvalData d);
```
please add a comment briefly explaining why we need prEvaluate.

mperego commented on src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp:
```diff
@@ -361,10 +361,13 @@ evaluateFields(typename Traits::EvalData workset)
   // Are we differentiating w.r.t. this parameter?
-  const bool is_active = (workset.dist_param_deriv_name == this->param_name) && (g_px_is_active||g_pp_is_active);
-  const bool is_direction_active = (workset.hessianWorkset.dist_param_deriv_direction_name == this->param_name) && (g_xp_is_active || g_pp_is_active);
+  const bool is_active = (workset.dist_param_deriv_name == this->param_name) && (g_px_is_active||g_pp_is_active||f_px_is_active||f_pp_is_active);
```
please add a comment briefly explaining what `is_active` and `is_direction_active` are

mperego commented on src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp:
```diff
@@ -439,10 +442,13 @@ evaluateFields(typename Traits::EvalData workset)
   // Are we differentiating w.r.t. this parameter?
-  const bool is_active = (workset.dist_param_deriv_name == this->param_name) && (g_px_is_active||g_pp_is_active);
-  const bool is_direction_active = (workset.hessianWorkset.dist_param_deriv_direction_name == this->param_name) && (g_xp_is_active || g_pp_is_active);
+  const bool is_active = (workset.dist_param_deriv_name == this->param_name) && (g_px_is_active||g_pp_is_active||f_px_is_active||f_pp_is_active);
```
same here, please add comment

mperego commented on src/evaluators/scatter/PHAL_ScatterResidual.hpp:
```diff
@@ -290,6 +290,7 @@ class ScatterResidualWithExtrudedParams<PHAL::AlbanyTraits::HessianVec,Traits>
                       PHX::FieldManager<Traits>& vm) {
     ScatterResidual<PHAL::AlbanyTraits::HessianVec, Traits>::postRegistrationSetup(d,vm);
   }
+  void evaluate2DFieldsDerivativesDueToExtrudedParams(typename Traits::EvalData d);
```
Is this used? (we can chat about that later)

mperego reviewed (commented):
Looks good to me. Thanks! 
I left a few comments. In addition to those, can you add a brief comment where you declare the  evaluators in the Gather and Scatter for the HessianVec type, briefly explaining what you are doing. Something similar to what you did in the text of this PR.

mperego commented:
@ikalash I mainly asked @kliegeois to add some comments. In the mean time I think you can test it, if you like to.

ikalash commented:
I checked this in a couple of my nightlies, and I think it is good to merge in.  Should issues crop up in some of the other nightlies, we can fix them later.

mperego commented:
> I checked this in a couple of my nightlies, and I think it is good to merge in. Should issues crop up in some of the other nightlies, we can fix them later.

OK, thanks for checking!

kliegeois commented:
I am currently working on the comments, I will push them today.

mperego commented:
@kliegeois one more thing. If you don't mind, can you please rebase and clean up a bit the commits?

kliegeois commented:
@mperego, I will do that, thanks!

mperego commented:
> @mperego, I will do that, thanks!

Thanks!

kliegeois commented:
@mperego , please do not merge this PR today, I observed an issue with one of my MPI test.
I am on it.

kliegeois commented:
@mperego , in the meantime, I rebased the branch to squash the commits and added the comments as requested.
I verified that those comments can be compiled with Doxygen, can you let me know if they are sufficiently clear to you?
Thanks a lot!

mperego reviewed (approved):
Comments are great. Thanks!

kliegeois commented:
@mperego , thanks for your review. I found the bug, I am correcting it. I test it and I let you know when the PR can be merged!

kliegeois commented:
@mperego , my last commit [3b649dc1c313519671f086778295ce422ed6acab](https://github.com/SNLComputation/Albany/pull/637/commits/3b649dc1c313519671f086778295ce422ed6acab) solved the issue.
The PR can be merged!
Thanks.

mperego commented:
> @mperego , my last commit [3b649dc1c313519671f086778295ce422ed6acab](https://github.com/SNLComputation/Albany/pull/637/commits/3b649dc1c313519671f086778295ce422ed6acab) solved the issue.
> The PR can be merged!
> Thanks.

Merged! Thanks.
